### PR TITLE
ADCS Edge Composition Visibility

### PR DESCRIPTION
**Pull request by elikmiller**
## Description

- Adds new endpoint `GET /api/v2/graphs/edge-composition` which, for edges of type `ADCSESC1` and `GoldenCert`, will return a graph representing the prerequisite nodes and edges which were used in the creation of the provided `ADCSESC1` or `GoldenCert` edge
- Updates the default node colors for nodes of type `AIACA`, `RootCA`, `EnterpriseCA`, `NTAuthStore` and `CertTemplate` to improve legibility
- Adds edge information panel help text for edges of type `ADCSESC1`
- Adds new `Composition` section to edge information panels for edges of type `ADCSESC1` and `GoldenCert`. Expanding this section in the UI will display the results of the new `GET /api/v2/graphs/edge-composition` in both list form and graph form
- Replaces the `faBuilding` icon used to represent AD Domains in the `DataSelector` component with `faGlobe` bringing it inline with the icon used to represent AD Domains elsewhere in the application
- Ensures new non-transit edges will not appear when making pathfinding queries via the UI
- Ensure new pathfinding edges are included in the existing suite of prebuilt cypher queries

## Motivation and Context

Currently, post-processed edges (such as `ADCSESC1`) can appear in the explore page graph. However, it may not be immediately obvious how this edge came into existence. This changeset seeks to help inform the user how specific post-processed edges are created so that they might better understand how to execute or remediate the relevant attack path.

## How Has This Been Tested?

Existing tests pass. No new tests have been included in this changeset.

## Screenshots

Screen Shot 2023-11-16 at 15 52 14

Screen Shot 2023-11-16 at 15 52 35

## Types of changes

-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [ ] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My changes include a database migration.

### CODE REVIEW

File: cmd/ui/src/views/QA/DataSelector/DataSelector.tsx [hunk 1]
```diff
@@ -159,7 +159,7 @@ const DataSelector: React.FC<{
                                     </Tooltip>
                                     <FontAwesomeIcon
                                         style={{ width: '10%', alignSelf: 'center' }}
-                                        icon={item.type === 'azure' ? faCloud : faBuilding}
+                                        icon={item.type === 'azure' ? faCloud : faGlobe}
```
Comment (urangel) [inline cmd/ui/src/views/QA/DataSelector/DataSelector.tsx:162]: Similar changes need to be made in the DomainsSelector component

File: cmd/api/src/api/registration/v2.go [hunk 2]
```diff
@@ -270,6 +270,9 @@ func NewV2API(cfg config.Configuration, resources v2.Resources, routerInst *rout
 		routerInst.GET(fmt.Sprintf("/api/v2/azure-tenants/{%s}/data-quality-stats", api.URIPathVariableTenantID), resources.GetAzureDataQualityStats).RequirePermissions(permissions.GraphDBRead),
 		routerInst.GET(fmt.Sprintf("/api/v2/platform/{%s}/data-quality-stats", api.URIPathVariablePlatformID), resources.GetPlatformAggregateStats).RequirePermissions(permissions.GraphDBRead),
 
+		// Edge composition API
+		routerInst.GET("/api/v2/graphs/edge-composition", resources.GetEdgeComposition).RequirePermissions(permissions.GraphDBRead).CheckFeatureFlag(resources.DB, appcfg.FeatureAdcs),
+
```
Comment (urangel) [inline cmd/api/src/api/registration/v2.go:275]: This is not too important but we should define this closer to the other api/v2/graphs/ endpoints

File: cmd/ui/src/views/Explore/EdgeInfo/EdgeInfoContent.tsx [hunk 3]
```diff
@@ -15,13 +15,19 @@
 // SPDX-License-Identifier: Apache-2.0
 
 import { Box, Divider, Typography, useTheme } from '@mui/material';
-import { EdgeInfoComponents, EdgeSections, SelectedEdge } from 'bh-shared-ui';
+import { EdgeInfoComponents, EdgeSections, SelectedEdge, apiClient } from 'bh-shared-ui';
+import isEmpty from 'lodash/isEmpty';
 import { FC, Fragment } from 'react';
+import { useDispatch } from 'react-redux';
```
Comment (urangel) [inline cmd/ui/src/views/Explore/EdgeInfo/EdgeInfoContent.tsx:21]: We should be importing useAppDispatch here instead

File: packages/go/analysis/ad/queries.go [hunk 4]
```diff
@@ -1407,6 +1407,22 @@ func FetchEnterpriseCAsCertChainPathToDomain(tx graph.Transaction, enterpriseCA,
...
+func FetchEnterpriseCAsTrustedForAuthPathToDomain(tx graph.Transaction, enterpriseCA, domain *graph.Node) (graph.PathSet, error) {
+	return ops.TraversePaths(tx, ops.TraversalPlan{
+		Root:      enterpriseCA,
+		Direction: graph.DirectionOutbound,
+		BranchQuery: func() graph.Criteria {
+			return query.KindIn(query.Relationship(), ad.TrustedForNTAuth, ad.NTAuthStoreFor)
+		},
+		//DescentFilter: func(ctx *ops.TraversalContext, segment *graph.PathSegment) bool {
+		//	return !segment.Trunk.Node.Kinds.ContainsOneOf(ad.Domain)
+		//},
```
Comment (urangel) [inline packages/go/analysis/ad/queries.go:1419]: This should be removed if it is not needed

File: packages/javascript/bh-shared-ui/src/components/HelpTexts/ADCSESC1/General.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import { groupSpecialFormat } from '../utils';
+import { EdgeInfoProps } from '../index';
+import { Typography } from '@mui/material';
+
+const General: FC<EdgeInfoProps> = ({ sourceName, sourceType, targetName }) => {
+    return (
+        <>
+            <Typography variant='body2'>
+                {groupSpecialFormat(sourceType, sourceName)} the privileges to perform the ADCS ESC1 attack against the
+                target domain.
```
Comment (urangel) [inline packages/javascript/bh-shared-ui/src/components/HelpTexts/ADCSESC1/General.tsx:27]: Is there a word missing here before 'the privileges'? Maybe this should be 'has the privileges...'

File: cmd/ui/src/views/QA/DataSelector/DataSelector.tsx (hunk 1 above)
Comment (urangel) [reply cmd/ui/src/views/QA/DataSelector/DataSelector.tsx:162]: nvm

File: packages/javascript/bh-shared-ui/src/components/HelpTexts/ADCSESC1/General.tsx (hunk 5 above)
Comment (elikmiller) [reply packages/javascript/bh-shared-ui/src/components/HelpTexts/ADCSESC1/General.tsx:27]: This is accounted for in the implementation of `groupSpecialFormat`

**Review by urangel — APPROVED** [review]

**Pull request merged**